File: tests/test_solenoid.c
/*  Confirm that the code accurately calculates the magnetic field at the 
    center of a current carrying wire loop. The loop is in the xy plane. 
    The observation points are on a line on the z-axis. The magnetic field
    should be only in the z-direction.

    Note: this test can be performed for either a loop or a solenoid. 
    TODO: make the solenoid a separate test. 
*/

#include <stdio.h>
#include <time.h>
#include <stdint.h>
#include <stdbool.h>
#include "../include/thor.h"

// Constant parameters for the problem
const double I = 1e3;           // Total current in loop 
const double R = 1.0/(2*PI);    // Radius of loop
double zspan = 5.0;            // Length of line to calculate results on
double compression = 2.0;       // 'compress' target pts at center of solenoid
// if the solenoid is requested, then the element size changes to 1 cm 
// and 100 elements/turn


// call this function like:
//  ./test_loop.o NUM_SOURCES NUM_TARGETS NUM_THREADS PHI
int main(int argc, char *argv[]) {

    // ---
    // Setup command line arguments
    // ---

    uint32_t n_sources, n_targets, n_threads;
    double phi;

    if (argc == 5) {
        n_sources = atoi(argv[1]);
        n_targets = atoi(argv[2]);
        n_threads = atoi(argv[3]);
        phi = (double)atof(argv[4]);
    }
    else {
        printf("Error. Incorrect number of arguments specified.\n");
        return -1;
    }

    // ---
    // Create target points along the z-axis and allocate results arrays
    // --- 

    // Compress points to region where analytical calculation is accurate
    double zcenter = zspan / 2.0; 
    double line_start = zcenter - ( zspan / (2.0 * compression) );
    double line_end = line_start + zspan / compression;

    Line *line_direct = new_line(Z, line_start, line_end, n_targets);
    Line *line_octree = new_line(Z, line_start, line_end, n_targets);
    double *Bz_analytical = calloc(n_targets, sizeof(double));


    // --- 
    // Create source points 
    // --- 

    double *xs = calloc(n_sources, sizeof(double));
    double *ys = calloc(n_sources, sizeof(double));
    double *zs = calloc(n_sources, sizeof(double));
    double *vol = calloc(n_sources, sizeof(double));
    double *Jx = calloc(n_sources, sizeof(double));
    double *Jy = calloc(n_sources, sizeof(double));
    double *Jz = calloc(n_sources, sizeof(double));
    // Need to do some extra work to place points in a spiral 
    double theta_step = 2*PI / (double)(n_sources - 1); 
    double step = theta_step*R;
    double side_length = R / 1000; 
    double area = side_length * side_length;
    double total_volume = area * 2*PI*R;
    double v = total_volume/n_sources;
    double J = I / area;
    double z = 0.0;
    double theta = 0; 
    double zstep = 0.0;

    int elements_per_turn = 100;
    side_length = zspan / ((double)n_sources / (double)elements_per_turn);
    area = side_length*side_length;
    v = side_length*area;        
    theta_step = 2.0*PI/(double)(elements_per_turn - 1);
    zstep = zspan/(double)n_sources; 
    J = I / area;
    

    for (size_t i = 0; i<n_sources; i++) {
        xs[i] = R*cos(theta); 
        ys[i] = R*sin(theta);
        zs[i] = z;
        vol[i] = v;
        Jx[i] = -J*sin(theta); 
        Jy[i] = J*cos(theta);
        theta += theta_step;
        z += zstep;
    }

    // ---
    // Compute solutions
    // ---

    // Analytical solution to test both direct and octree methods against
    // Should not change along the axis
    double Bz_solenoid = bfield_ideal_solenoid(100.0, I); 
    for (size_t i=0; i<n_targets; i++) {
        Bz_analytical[i] = Bz_solenoid;
    }

    // Octree calculation (timed internally but repeated here for calculation of speedup)
    time_t start = clock();
    bfield_octree(
        xs, ys, zs, vol, Jx, Jy, Jz, n_sources, 
        line_octree->x, line_octree->y, line_octree->z, n_targets, 
        line_octree->Bx, line_octree->By, line_octree->Bz, 1, phi);
    time_t end = clock();
    double time_octree = (double)(end-start)/CLOCKS_PER_SEC;
    printf("\n");
    printf("Octree calculation time: %f s\n", time_octree);

    // Direct summation of all points
    start = clock();
    bfield_direct(
        xs, ys, zs, vol, Jx, Jy, Jz, n_sources, 
        line_direct->x, line_direct->y, line_direct->z, n_targets, 
        line_direct->Bx, line_direct->By, line_direct->Bz, 1);
    end = clock(); 
    double time_direct = (double)(end-start)/CLOCKS_PER_SEC;
    printf("Direct calculation time: %f s\n", time_direct);
    printf("Speedup: %.3fx\n", time_direct/time_octree);
    printf("\n");
    
    // --- 
    // Compare solutions 
    // --- 

    double Bz_direct_error = rms_error(line_direct->Bz, Bz_analytical, n_targets); 
    double Bz_octree_error = rms_error(line_octree->Bz, Bz_analytical, n_targets);
    double octree_direct_error = rms_error(line_octree->Bz, line_direct->Bz, n_targets);

    printf("RMS errors relative to analytical:\n");
    printf("Bz  direct error: %.3f %%\n", 100*Bz_direct_error);
    printf("Bz  octree error: %.3f %%\n", 100*Bz_octree_error);
    printf("\n");
    printf("RMS error relative to direct:\n");
    printf("Bz octree error: %.3f %%\n", 100*octree_direct_error);

    int j = n_targets/2;
    printf("Analytical field at z = %.3f: %.3f T\n", line_direct->z[j], Bz_analytical[j]);
    printf("Direct     field at z = %.3f: %.3f T\n", line_direct->z[j], line_direct->z[j]);
    printf("Octree     field at z = %.3f: %.3f T\n", line_direct->z[j], line_octree->z[j]);


    // Ctrl+F gives 8 `calloc()` and 8 `free()` commands in this function
    free(Bz_analytical); 
    free_line(line_direct);
    free_line(line_octree);
    free(xs); free(ys); free(zs); free(vol); free(Jx); free(Jy); free(Jz);
}

File: tests/jl/test_solenoid.jl
""" Test the octree code on the solenoid problem 
"""

cd(@__DIR__)
include("../../src/thor.jl")
using Plots, Printf

# Problem parameters
const MU0_OVER_4PI = 1e-7 
const MU0 = 4*pi *MU0_OVER_4PI 
const h = 1e-2                      # [m], element size
const n = 1/h                       # [turns/m] and [elements/turn]
const I = 1000                      # [A] 
const r = 1/(2pi)                   # [m], solenoid radius 
const L = 20*r                      # [m], solenoid length
const Bz_expected = MU0 * n * I     # [T], Bz 

# Setup functions 

# Mesh a solenoid given the problem parameters 
function make_solenoid(N::Int64) 

    if N < 32000 
        N = 32000 
    end

    pts = zeros(N,3) 
    vol = (h^3) .* ones(N) 
    J = zeros(N,3) 
    m = convert(Int64, ceil(n))     # shadows `n` above
    area = h*h 
    j = I/area

    if N % m != 0 
        N = ceil(N/m) * m 
    end

    theta_step = 2*pi/n
    zstep = h 
    z = L/2 

    # Start from z=0 and mesh each turn of the solenoid individually
    k = 1 
    for _ in 2:N/m 
        theta = 0 
        for _ in 1:m 
            pts[k,1] = r*cos(theta) 
            pts[k,2] = r*sin(theta)
            pts[k,3] = z
            J[k,1] = -j*sin(theta) 
            J[k,2] = j*cos(theta)
            theta += theta_step 
            k+=1
        end
        # z += zstep 
    end

    return pts, vol, J
end

function make_target_pts(M::Int64) 
    target_pts = zeros(M,3)
    Lstep = L / (M - 1.0) 
    for i in 2:M 
        target_pts[i,3] = target_pts[i-1,3] + Lstep 
    end 

    return target_pts
end

# Measure the RMS error of y against known x 
# Note: ignores x[i] < 1e-8
function rms_error(y::Vector{<:Real}, x::Vector{<:Real})
    err = 0.0
    if length(y) != length(x) 
        @printf "Error in rms_error(): y and x are not the same length!\n"
        return 1e32
    end
    for i in 1:length(y) 
        if x[i] > 1e-8 
            err += ((y[i] - x[i]) / x[i])^2
        end
    end
    return sqrt(err/length(y))
end

N = 20000 
M = 1000
phi = 1e-1

source_pts, vol, J = make_solenoid(N) 
target_pts = make_target_pts(M) 

Bexact = MU0 * n * I 
# time_octree = @elapsed begin 
    Boctree = bfield(source_pts, vol, J, target_pts, octree; phi=phi)
# end
# time_direct = @elapsed begin 
    Bdirect = bfield(source_pts, vol, J, target_pts, direct) 
# end

# Errors at the midpoint (to compare against analytical)
i = convert(Int64, floor(M/2.0))
rms_error_octree = (Boctree[i,3] - Bexact) / Bexact
rms_error_direct = (Bdirect[i,3] - Bexact) / Bexact
@printf "Comparing octree and direct methods to exact solution: \n---\n"
@printf "Expected value: %.3f T \n" Bexact 
@printf "Octree result:  %.3f T (%.2f %% error)\n" Boctree[i,3] rms_error_octree*100.0
@printf "Direct result:  %.3f T (%.2f %% error)\n" Bdirect[i,3] rms_error_direct*100.0
@printf "\n"
@printf "Comparing octree to direct method: \n---\n"
rms_error_octree = rms_error(Boctree[:,3], Bdirect[:,3])
@printf "Octree MRS error relative to direct: %.3f %% \n" rms_error_octree*100.0

@printf "\n"
@printf "Time for direct solution: %.3f s\n" time_direct 
@printf "Time for octree solution: %.3f s\n" time_octree 
@printf "Octree speedup: %.1fx \n" time_direct/time_octree

p = plot()
plot!(p, target_pts[:,3], Boctree[:,3], label="octree")
plot!(p, target_pts[:,3], Bdirect[:,3], label="direct", linestyle=:dash)
xlabel!(p, "Distance Along Solenoid Axis [m]")
ylabel!(p, "Magnetic Flux Density (Bz) [T]")
title!(p, "Solenoid Test: Octree vs Direct Integration (\$\\phi\$="*string(phi)*")")



display(p)

# p2 = plot() 
# scatter!(source_pts[:,1], source_pts[:,2], source_pts[:,3])

